connection_producer.go

```diff
@@ -0,0 +1,218 @@
+	}
+
+	var db *sql.DB
+	var err error
+	if c.PrivateKey != "" {
+		db, err = openSnowflake(c.ConnectionURL, c.Username, c.PrivateKey)
+		if err != nil {
+			return nil, fmt.Errorf("error opening Snowflake connection: %w", err)
+		}
+	} else {
+		db, err = sql.Open(snowflakeSQLTypeName, c.ConnectionURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open connection: %w", err)
+		}
+	}
+
+	c.snowflakeDB = db
+
+	return db, nil
+}
+
+// close terminates the database connection without locking
+func (c *snowflakeConnectionProducer) close() error {
+	if c.snowflakeDB != nil {
+		if err := c.snowflakeDB.Close(); err != nil {
+			return err
+		}
+	}
+
+	c.snowflakeDB = nil
+	return nil
+}
+
+// Close terminates the database connection with locking
+func (c *snowflakeConnectionProducer) Close() error {
+	c.Lock()
+	defer c.Unlock()
+
+	return c.close()
+}
+
+// Open the DB connection to Snowflake or return an error.
+func openSnowflake(connectionURL, username, providedPrivateKey string) (*sql.DB, error) {
+	// Parse thee connection_url for required fields. Should be of
+	// the form <account_name>.snowflakecomputing.com/<db_name>
+	accountName, dbName, err := parseSnowflakeFieldsFromURL(connectionURL)
+	if err != nil {
+		return nil, err
+	}
+
+	privateKey, err := getPrivateKey(providedPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	snowflakeConfig := &gosnowflake.Config{
+		Account:       accountName,
+		Database:      dbName,
+		User:          username,
+		Authenticator: gosnowflake.AuthTypeJwt,
+		PrivateKey:    privateKey,
+	}
+	connector := gosnowflake.NewConnector(gosnowflake.SnowflakeDriver{}, *snowflakeConfig)
+
+	return sql.OpenDB(connector), nil
+}
+
+// parseSnowflakeFieldsFromURL uses a regex to extract account and DB
+// info from a connectionURL
+func parseSnowflakeFieldsFromURL(connectionURL string) (string, string, error) {
+	if !accountAndDBNameFromConnURLRegex.MatchString(connectionURL) {
+		return "", "", ErrInvalidSnowflakeURL
+	}
+	res := accountAndDBNameFromConnURLRegex.FindStringSubmatch(connectionURL)
+	if len(res) != 3 {
+		return "", "", ErrInvalidSnowflakeURL
+	}
+
+	return res[1], res[2], nil
+}
+
+// Open and decode the private key file
+func getPrivateKey(providedPrivateKey string) (*rsa.PrivateKey, error) {
+	var block *pem.Block
+
+	// Try loading a file with the provided private key field first. If the the file doesn't
+	// exist, assume they provided the raw key and decode it. Otherwise return an error. If there
+	// was no error, then they likely provided a file path to a private key.
+	keyFile, err := os.ReadFile(providedPrivateKey)
+	if err != nil {
+		if os.IsNotExist(err) {
+			block, _ = pem.Decode([]byte(providedPrivateKey))
+		} else {
+			return nil, fmt.Errorf("failed to read private key file: %w", err)
+		}
+	} else {
+		block, _ = pem.Decode(keyFile)
+	}
+
+	if block == nil || block.Type != "PRIVATE KEY" {
+		return nil, fmt.Errorf("failed to decode the private key value")
+	}
```

Comment:
Should we split this into two separate conditionals and error responses? I image we may want to know if we were able to parse the block Type but it didn't match our expected value.